### PR TITLE
Revisit `torch._six.string_classes` removal (#94709)

### PR DESCRIPTION
**Pull request by XuehaiPan**
Revisit `torch._six.string_classes` (which is `(str, bytes)`) removal: `isinstance(obj, string_classes) -> isinstance(obj, str)`.

Both `str` and `bytes` are `Sequence` classes.

```python
In [1]: from typing import Sequence

In [2]: issubclass(bytes, Sequence)
Out[2]: True

In [3]: issubclass(str, Sequence)
Out[3]: True
```

Re-add `bytes` to type guards like:

```python
def is_seq(obj):
    return isinstance(obj, Sequence) and not isinstance(obj, (str, bytes))
```

Ref:

- https://github.com/pytorch/pytorch/pull/94709#issuecomment-1487282912
- #97737
- #97789

### CODE REVIEW

File: torch/onnx/_internal/jit_utils.py [hunk 1]
```diff
@@ -310,7 +310,7 @@ def _create_node(
 @_beartype.beartype
 def _is_onnx_list(value):
     return (
-        not isinstance(value, str)
+        not isinstance(value, (str, bytes))
         and not isinstance(value, torch.Tensor)
```
Comment (Skylion007) [inline torch/onnx/_internal/jit_utils.py:314]:
```suggestion
        not isinstance(value, (str, bytes, torch.Tensor))
```

**Review by Skylion007 — APPROVED** [review]

**Review by albanD — COMMENTED** [review]
This is not covering the change to:
- `torch/__init__,py`
- distributed_c10d.py
- rendezvous.py
- _serialization.py
- serialization.py
- module.py
- common_utils.py

Any reason why not?

**Comment by XuehaiPan** [comment]
> 

@albanD I have checked every line change in the original PR. I think `bytes` are hardly ever used in these cases. Users are using `str` instead.

```python
```

**Comment by XuehaiPan** [comment]
Closed by mistake.

**Comment by albanD** [comment]
> I think bytes are hardly ever used in these cases.

That is not a great justification!  bytes are hardly used in pinned memory but it leads to very bad perf for users in practice.
So we should be leaning the other way: make sure they are exactly the same and clearly know what is the impact of any change.

If we don't want to dive into details on how each of these APIs is called across all use cases on github (which I don't think we should do as it will take quite a bit of time) we should just update all of these back.

**Comment by XuehaiPan** [comment]
> So we should be leaning the other way: make sure they are exactly the same and clearly know what is the impact of any change.

I'll give a list of changes in detail tomorrow.

**Comment by albanD** [comment]
@XuehaiPan sounds good. Thanks for taking the time to fix it up!

**Comment by XuehaiPan** [comment]
## Full list of `torch._six.string_classes` changes

Items filled with `X` markup are reverted. The remaining items are followed with reasons why they are not changed back. Feel free to ask for further changes if the listed reasons are not reasonable.

-----

- [X] `test/distributed/test_store.py` (reverted in #94709)

<details>
<summary>Code</summary>

```python
class MyPythonStore(dist.Store):
    def __init__(self):
        super().__init__()
        self.store = {}

    def set(self, key, value):
        if not isinstance(key, (str, bytes)):  # <=== reverted
            raise AssertionError("Expected set to be called with string key")
        if type(value) is not bytes:
            raise AssertionError("Expected set to be called with bytes value")
        self.store[key] = value
    def get(self, key):
        value = self.store.get(key, b"")
        if type(value) is not bytes:
            raise AssertionError("Expected get to return bytes value")
        return value
    def add(self, key, value):
        new = int(self.store.get(key, 0)) + value
        self.set(key, bytes(str(new).encode("utf-8")))
        return new
```

</details>

------

- [ ] `test/test_torch.py`

Reason for not to re-add `bytes`:

1. This function has not been called. It is defined but not referenced.
2. Skipped namespaces should be `str`s or `re.Pattern`s, rather than `bytes`s.

<details>
<summary>Code</summary>

```python
    def test_doc(self):
        checked_types = (types.MethodType, types.FunctionType,
                         types.BuiltinFunctionType, types.BuiltinMethodType)
        def _test_namespace(ns, *skips):
            if isinstance(ns, object):
                ns_name = ns.__class__.__name__
            else:
                ns_name = ns.__name__
            skip_regexes = []
            for r in skips:
                if isinstance(r, str):  # <=== leave as `str`-only
                    skip_regexes.append(re.compile('^{}$'.format(re.escape(r))))
                else:
                    skip_regexes.append(r)

            ...
```

</details>

------

- [ ] `torch/__init__.py`

Reason for not to re-add `bytes`:

1. `torch.set_default_tensor_type(dtype)` mostly is called by `str` or `torch.dtype`s.
2. In docstrings, the arguments should be `type or string`.
3. `__import_dotted_name` only accepts `str`.

<details>
<summary>Code</summary>

```python
def set_default_tensor_type(t):
    r"""Sets the default ``torch.Tensor`` type to floating point tensor type
    ``t``. This type will also be used as default floating point type for
    type inference in :func:`torch.tensor`.
    The default floating point tensor type is initially ``torch.FloatTensor``.
    Args:
        t (type or string): the floating point tensor type or its name
    Example::
        >>> # xdoctest: +SKIP("Other tests may have changed the default type. Can we reset it?")
        >>> torch.tensor([1.2, 3]).dtype    # initial default for floating point is torch.float32
        torch.float32
        >>> torch.set_default_tensor_type(torch.DoubleTensor)
        >>> torch.tensor([1.2, 3]).dtype    # a new floating point tensor
        torch.float64
    """
    if isinstance(t, str):  # <=== leave as `str`-only
        t = _import_dotted_name(t)
    _C._set_default_tensor_type(t)
```

</details>

------

- [X] `torch/cuda/amp/autocast_mode.py` (reverted in #94709)

<details>
<summary>Code</summary>

```python
# Casts Tensors and containers of Tensors.  Special-cases passthroughs for strings and np.ndarrays, which
# may be falsely detected as "Iterables."
def _cast(value, dtype):
    if isinstance(value, torch.Tensor):
        is_eligible = (value.is_floating_point() and value.is_cuda and (value.dtype is not torch.float64))
        return value.to(dtype) if is_eligible else value
    elif isinstance(value, (str, bytes)):  # <=== reverted
        return value
    elif HAS_NUMPY and isinstance(value, np.ndarray):
        return value
    elif isinstance(value, collections.abc.Mapping):
        return {_cast(k, dtype): _cast(v, dtype) for k, v in value.items()}
    elif isinstance(value, collections.abc.Iterable):
        iterable = map(lambda v: _cast(v, dtype), value)
        if isinstance(value, (list, tuple)):
            return type(value)(iterable)
        else:
            return iterable
    else:
        return value
```

</details>

------

- [ ] `torch/distributed/distributed_c10d.py`

Reason for not to re-add `bytes`:

1. The type annotation shows `name` should be a `str`.
2. `getattr(obj, name, default)` only accepts `str`.

<details>
<summary>Code</summary>

```python
    def __new__(cls, name: str):
        if not isinstance(name, str):  # <=== leave as `str`-only
            raise ValueError("Backend name must be a string, but got: {}".format(name))
        value = getattr(Backend, name.upper(), Backend.UNDEFINED)

        if value != Backend.GLOO and value != Backend.NCCL and value != Backend.UCC and value != Backend.MPI:
            value = name.lower()
        return value
```

</details>

------

- [ ] `torch/distributed/rendezvous.py`

Reason for not to re-add `bytes`:

1. The type annotation shows `url` should be a `str`.
2. `url` are mostly passed as `str`.

<details>
<summary>Code</summary>

```python
def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
    if not isinstance(url, str):  # <=== leave as `str`-only
        raise RuntimeError("`url` must be a string. {}: {}".format(type(url), url))

    if not isinstance(rank, numbers.Integral):
        raise RuntimeError("`rank` must be an integer. {}".format(rank))

    ...
```

</details>

------

- [ ] `torch/jit/_serialization.py`

Reason for not to re-add `bytes`:

1. In docstrings, the arguments should be `file-like or string`.
2. If the argument `f` is a file path, it should be a `str` rather than `bytes`.

<details>
<summary>Code</summary>

```python
def load(f, map_location=None, _extra_files=None, _restore_shapes=False):
    r"""
    Load a :class:`ScriptModule` or :class:`ScriptFunction` previously
    saved with :func:`torch.jit.save <torch.jit.save>`
    All previously saved modules, no matter their device, are first loaded onto CPU,
    and then are moved to the devices they were saved from. If this fails (e.g.
    because the run time system doesn't have certain devices), an exception is
    raised.
    Args:
        f: a file-like object (has to implement read, readline, tell, and seek),
            or a string containing a file name
        map_location (string or torch.device): A simplified version of
            ``map_location`` in `torch.jit.save` used to dynamically remap
            storages to an alternative set of devices.
        _extra_files (dictionary of filename to content): The extra
            filenames given in the map would be loaded and their content
            would be stored in the provided map.
        _restore_shapes (bool): Whether or not to retrace the module on load using stored inputs
    Returns:
        A :class:`ScriptModule` object.
    Example:
    .. testcode::
        import torch
        import io
        torch.jit.load('scriptmodule.pt')
        # Load ScriptModule from io.BytesIO object
        with open('scriptmodule.pt', 'rb') as f:
            buffer = io.BytesIO(f.read())
        # Load all tensors to the original device
        torch.jit.load(buffer)
        # Load all tensors onto CPU, using a device
        buffer.seek(0)
        torch.jit.load(buffer, map_location=torch.device('cpu'))
        # Load all tensors onto CPU, using a string
        buffer.seek(0)
        torch.jit.load(buffer, map_location='cpu')
        # Load with extra files.
        extra_files = {'foo.txt': ''}  # values will be replaced with data
        torch.jit.load('scriptmodule.pt', _extra_files=extra_files)
        print(extra_files['foo.txt'])
    .. testoutput::
        :hide:
        ...
    .. testcleanup::
        import os
        os.remove("scriptmodule.pt")
    """

    if isinstance(f, str):  # <=== leave as `str`-only
        if not os.path.exists(f):  # type: ignore[type-var]
            raise ValueError("The provided filename {} does not exist".format(f))  # type: ignore[str-bytes-safe]
        if os.path.isdir(f):
            raise ValueError("The provided filename {} is a directory".format(f))  # type: ignore[str-bytes-safe]
    map_location = validate_map_location(map_location)
    if _extra_files is None:
        _extra_files = {}
    
    ...
```

</details>

------

- [ ] `torch/nn/modules/module.py`

Reason for not to re-add `bytes`:

1. The type annotation shows `name` should be a `str`.
2. If we accept `bytes` as name, it will break `getattr(module, name)` and `module.name`.

<details>
<summary>Code</summary>

```python
    def register_buffer(self, name: str, tensor: Optional[Tensor], persistent: bool = True) -> None:
        r"""Adds a buffer to the module.
        This is typically used to register a buffer that should not to be
        considered a model parameter. For example, BatchNorm's ``running_mean``
        is not a parameter, but is part of the module's state. Buffers, by
        default, are persistent and will be saved alongside parameters. This
        behavior can be changed by setting :attr:`persistent` to ``False``. The
        only difference between a persistent buffer and a non-persistent buffer
        is that the latter will not be a part of this module's
        :attr:`state_dict`.
        Buffers can be accessed as attributes using given names.
        Args:
            name (str): name of the buffer. The buffer can be accessed
                from this module using the given name
            tensor (Tensor or None): buffer to be registered. If ``None``, then operations
                that run on buffers, such as :attr:`cuda`, are ignored. If ``None``,
                the buffer is **not** included in the module's :attr:`state_dict`.
            persistent (bool): whether the buffer is part of this module's
                :attr:`state_dict`.
        Example::
            >>> # xdoctest: +SKIP("undefined vars")
            >>> self.register_buffer('running_mean', torch.zeros(num_features))
        """
        if persistent is False and isinstance(self, torch.jit.ScriptModule):
            raise RuntimeError("ScriptModule does not support non-persistent buffers")
        if '_buffers' not in self.__dict__:
            raise AttributeError(
                "cannot assign buffer before Module.__init__() call")
        elif not isinstance(name, str):  # <=== leave as `str`-only
            raise TypeError("buffer name should be a string. "
                            "Got {}".format(torch.typename(name)))
        elif '.' in name:
            ...

        ...

    def register_parameter(self, name: str, param: Optional[Parameter]) -> None:
        r"""Adds a parameter to the module.
        The parameter can be accessed as an attribute using given name.
        Args:
            name (str): name of the parameter. The parameter can be accessed
                from this module using the given name
            param (Parameter or None): parameter to be added to the module. If
                ``None``, then operations that run on parameters, such as :attr:`cuda`,
                are ignored. If ``None``, the parameter is **not** included in the
                module's :attr:`state_dict`.
        """
        if '_parameters' not in self.__dict__:
            raise AttributeError(
                "cannot assign parameter before Module.__init__() call")

        elif not isinstance(name, str):  # <=== leave as `str`-only
            raise TypeError("parameter name should be a string. "
                            "Got {}".format(torch.typename(name)))
        elif '.' in name:
            ...

        ...

    def add_module(self, name: str, module: Optional['Module']) -> None:
        r"""Adds a child module to the current module.
        The module can be accessed as an attribute using the given name.
        Args:
            name (str): name of the child module. The child module can be
                accessed from this module using the given name
            module (Module): child module to be added to the module.
        """
        if not isinstance(module, Module) and module is not None:
            raise TypeError("{} is not a Module subclass".format(
                torch.typename(module)))
        elif not isinstance(name, str):  # <=== leave as `str`-only
            raise TypeError("module name should be a string. Got {}".format(
                torch.typename(name)))
        elif hasattr(self, name) and name not in self._modules:
            ...

        ...
```

</details>

------

- [X] `torch/onnx/_internal/jit_utils.py` (reverted in #94709)

<details>
<summary>Code</summary>

```python
@_beartype.beartype
def _is_onnx_list(value):
    return (
        not isinstance(value, (str, bytes))  # <=== reverted
        and not isinstance(value, torch.Tensor)
        and isinstance(value, Iterable)
    )
```

</details>

------

- [ ] `torch/serialization.py`

Reason for not to re-add `bytes`:

1. If the argument `map_location` is a file path, it should be a `str` rather than `bytes`.

<details>
<summary>Code</summary>

```python
def _get_restore_location(map_location):
    if map_location is None:
        restore_location = default_restore_location
    elif isinstance(map_location, dict):
        def restore_location(storage, location):
            location = map_location.get(location, location)
            return default_restore_location(storage, location)
    elif isinstance(map_location, str):  # <=== leave as `str`-only
        def restore_location(storage, location):
            return default_restore_location(storage, map_location)
    elif isinstance(map_location, torch.device):
        def restore_location(storage, location):
            return default_restore_location(storage, str(map_location))
    else:
        def restore_location(storage, location):
            result = map_location(storage, location)
            if result is None:
                result = default_restore_location(storage, location)
            return result
    return restore_location
```

</details>

------

- [ ] `torch/testing/_internal/common_utils.py`

Reason for not to re-add `bytes`:

1. Internal use and the argument is never passed with `bytes`.

<details>
<summary>Code</summary>

```python
def shell(command, cwd=None, env=None, stdout=None, stderr=None):
    sys.stdout.flush()
    sys.stderr.flush()
    # The following cool snippet is copied from Py3 core library subprocess.call
    # only the with
    #   1. `except KeyboardInterrupt` block added for SIGINT handling.
    #   2. In Py2, subprocess.Popen doesn't return a context manager, so we do
    #      `p.wait()` in a `final` block for the code to be portable.
    #
    # https://github.com/python/cpython/blob/71b6c1af727fbe13525fb734568057d78cea33f3/Lib/subprocess.py#L309-L323
    assert not isinstance(command, str), "Command to shell should be a list or tuple of tokens"  # <=== leave as `str`-only
    p = subprocess.Popen(command, universal_newlines=True, cwd=cwd, env=env, stdout=stdout, stderr=stderr)
    return wait_for_process(p)

class StringPair(UnittestPair):
    CLS = (str, bytes)  # <=== reverted
    TYPE_NAME = "string"
```

</details>

------

- [X] `torch/onnx/_internal/jit_utils.py` (reverted in #97789)

<details>
<summary>Code</summary>

```python
def default_convert(data):
    r"""
        Function that converts each NumPy array element into a :class:`torch.Tensor`. If the input is a `Sequence`,
        `Collection`, or `Mapping`, it tries to convert each element inside to a :class:`torch.Tensor`.
        If the input is not an NumPy array, it is left unchanged.
        This is used as the default function for collation when both `batch_sampler` and
        `batch_size` are NOT defined in :class:`~torch.utils.data.DataLoader`.
        The general input type to output type mapping is similar to that
        of :func:`~torch.utils.data.default_collate`. See the description there for more details.
        Args:
            data: a single data point to be converted
        Examples:
            >>> # xdoctest: +SKIP
            >>> # Example with `int`
            >>> default_convert(0)
            0
            >>> # Example with NumPy array
            >>> default_convert(np.array([0, 1]))
            tensor([0, 1])
            >>> # Example with NamedTuple
            >>> Point = namedtuple('Point', ['x', 'y'])
            >>> default_convert(Point(0, 0))
            Point(x=0, y=0)
            >>> default_convert(Point(np.array(0), np.array(0)))
            Point(x=tensor(0), y=tensor(0))
            >>> # Example with List
            >>> default_convert([np.array([0, 1]), np.array([2, 3])])
            [tensor([0, 1]), tensor([2, 3])]
    """
    elem_type = type(data)
    if isinstance(data, torch.Tensor):
        return data
    elif elem_type.__module__ == 'numpy' and elem_type.__name__ != 'str_' \
            and elem_type.__name__ != 'string_':
        # array of string classes and object
        if elem_type.__name__ == 'ndarray' \
                and np_str_obj_array_pattern.search(data.dtype.str) is not None:
            return data
        return torch.as_tensor(data)
    elif isinstance(data, collections.abc.Mapping):
        try:
            return elem_type({key: default_convert(data[key]) for key in data})
        except TypeError:
            # The mapping type may not support `__init__(iterable)`.
            return {key: default_convert(data[key]) for key in data}
    elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
        return elem_type(*(default_convert(d) for d in data))
    elif isinstance(data, tuple):
        return [default_convert(d) for d in data]  # Backwards compatibility.
    elif isinstance(data, collections.abc.Sequence) and not isinstance(data, (str, bytes)):  # <=== reverted
        try:
            return elem_type([default_convert(d) for d in data])
        except TypeError:
            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
            return [default_convert(d) for d in data]
    else:
        return data

default_collate_fn_map[str] = collate_str_fn
default_collate_fn_map[bytes] = collate_str_fn  # <=== reverted
```

</details>

------

- [X] `torch/utils/data/_utils/pin_memory.py` (reverted in #97737)

<details>
<summary>Code</summary>

```python
def pin_memory(data, device=None):
    if isinstance(data, torch.Tensor):
        return data.pin_memory(device)
    elif isinstance(data, (str, bytes)):  # <=== reverted
        return data
    elif isinstance(data, collections.abc.Mapping):
        try:
            return type(data)({k: pin_memory(sample, device) for k, sample in data.items()})  # type: ignore[call-arg]
        except TypeError:
            # The mapping type may not support `__init__(iterable)`.
            return {k: pin_memory(sample, device) for k, sample in data.items()}
    elif isinstance(data, tuple) and hasattr(data, '_fields'):  # namedtuple
        return type(data)(*(pin_memory(sample, device) for sample in data))
    elif isinstance(data, tuple):
        return [pin_memory(sample, device) for sample in data]  # Backwards compatibility.
    elif isinstance(data, collections.abc.Sequence):
        try:
            return type(data)([pin_memory(sample, device) for sample in data])  # type: ignore[call-arg]
        except TypeError:
            # The sequence type may not support `__init__(iterable)` (e.g., `range`).
            return [pin_memory(sample, device) for sample in data]
    elif hasattr(data, "pin_memory"):
        return data.pin_memory()
    else:
        return data
```

</details>

------

- [ ] `torch/utils/data/dataloader.py`

Reason for not to re-add `bytes`:

1. Passing by `bytes` always raises an exception.

<details>
<summary>Code</summary>

```python
    @multiprocessing_context.setter
    def multiprocessing_context(self, multiprocessing_context):
        if multiprocessing_context is not None:
            if self.num_workers > 0:
                if isinstance(multiprocessing_context, str):  # <=== leave as `str`-only
                    valid_start_methods = multiprocessing.get_all_start_methods()
                    if multiprocessing_context not in valid_start_methods:
                        raise ValueError(
                            ('multiprocessing_context option '
                             'should specify a valid start method in {!r}, but got '
                             'multiprocessing_context={!r}').format(valid_start_methods, multiprocessing_context))
                    # error: Argument 1 to "get_context" has incompatible type "Union[str, bytes]"; expected "str"  [arg-type]
                    multiprocessing_context = multiprocessing.get_context(multiprocessing_context)  # type: ignore[arg-type]
                if not isinstance(multiprocessing_context, python_multiprocessing.context.BaseContext):
                    raise TypeError(('multiprocessing_context option should be a valid context '
                                     'object or a string specifying the start method, but got '
                                     'multiprocessing_context={}').format(multiprocessing_context))
```

</details>

**Comment by albanD** [comment]
Thanks for the detailed study!
Some high level comments:
- "url are mostly passed as str." this is not a justification within PT I'm afraid! We have enough users and dependencies that we cannot assume that people won't do bad stuff!  Maybe someone had to re-encode their url and ended up with bytes instead of a str without knowing it.
- "If the argument map_location is a file path, it should be a str rather than bytes." this kind of things are based on type hints. But they are not really enforced. And more generally, our type hints are lacking/inacurate enough that I don't think our users expect them to be really accurate. So we shouldn't use type hints as a justification.
- "In docstrings, the arguments should be file-like or string." Well, given six, it looks like "string" is interpreted as both "str" and "bytes".
- I definitely agree with you that no one should be passing bytes to these functions. But there are so many uses of these APIs that, in my experience, if you can do something weird, someone will always rely on it: https://xkcd.com/1172/


The ones that raise errors when passed bytes are fine to leave out, but I think we should revert the others that are user facing.
- rendezvous.py should be changed back. urllib.parse.urlparse used there supports bytes properly.
- serialization.py there are too many callsite for me to easily figure out if this can be ok to be bytes or not. I would revert.

**Review by NivekT — COMMENTED** [review]
The DataLoader part makes sense to me. Might want to add this to 2.0.1 milestone?

**Review by albanD — APPROVED** [review]
Thanks!

File: torch/distributed/rendezvous.py [hunk 2]
```diff
@@ -90,7 +90,7 @@ def _rendezvous_helper(url: str, rank: int, world_size_opt: Optional[int], **kwa
 
 
 def rendezvous(url: str, rank: int = -1, world_size: int = -1, **kwargs):
-    if not isinstance(url, str):
+    if not isinstance(url, (str, bytes)):
         raise RuntimeError("`url` must be a string. {}: {}".format(type(url), url))
```
Comment (Skylion007) [inline torch/distributed/rendezvous.py:94]: This should probably be a TypeError. Also the functions signature of rendezvous is now wrong (url should be a Union[str, bytes] right?)

Comment (albanD) [reply torch/distributed/rendezvous.py:94]: Yes, can you open an issue so that we can ask the distributed team what they want to do about that?

**Pull request closed without merging**